sympy/combinatorics/fp_groups.py

```diff
@@ -688,4 +807,374 @@ def coset_enumeration_c(fp_grp, Y):
     return C
 
 
+def low_index_subgroups(G, N):
+    """
+    Implements the Low Index Subgroups algorithm, i.e find all subgroups of
+    "G" upto a given index "N". This implements the method described in
+    [Sim94]. This procedure involves a backtrack search over incomplete Coset
+    Tables, rather than over forced coincidences.
+
+    G: An FpGroup < X|R >
+    N: positive integer, representing the maximun index value for subgroups
+
+    References
+    ==========
+
+    [1] Holt, D., Eick, B., O'Brien, E.
+    "Handbook of Computational Group Theory"
+    Section 5.4
+
+    [2] Marston Conder and Peter Dobcsanyi
+    "Applications and Adaptions of the Low Index Subgroups Procedure"
+
+    """
+    C = CosetTable(G, [])
+    R = G.relators()
+    # length chosen for the length of the short relators
+    len_short_rel = 5
+    # elements of R2 only checked at the last step for complete
+    # coset tables
+    R2 = set([rel for rel in R if len(rel) > len_short_rel])
+    # elements of R1 are used in inner parts of the process to prune
+    # branches of the search tree,
+    R1 = set([rel.identity_cyclic_reduction() for rel in set(R) - R2])
+    R1_c_list = C.conjugates(R1)
+    S = []
+    for x in C.A:
+        descendant_subgroups(S, C, R1_c_list, x, R2, N)
+    return S
+
+
+def descendant_subgroups(S, C, R1_c_list, x, R2, N):
+    A_dict = C.A_dict
+    A_dict_inv = C.A_dict_inv
+    if C.is_complete():
+        # check whether the relators in R2 are satisfied
+        for w, alpha in product(R2, C.omega):
+            if not C.scan_check(alpha, w):
+                return
+        # relators in R2 are satisfied, append the table to list
+        S.append(C)
+    else:
+        for alpha, x in product(range(len(C.table)), C.A):
+            if C.table[alpha][A_dict[x]] is None:
+                # this is "x" in pseudo-code (using "y" makes it clear)
+                undefined_coset, undefined_gen = alpha, x
+                break
+        reach = C.omega + [C.n]
+        for beta in reach:
+            if beta == C.n:
+                C.table.append([None]*len(C.A))
+                C.p.append(len(C.A) - 1)
+            if beta < N and C.table[beta][A_dict_inv[undefined_gen]] is None:
+                try_descendant(S, C, R1_c_list, R2, N, undefined_coset, undefined_gen, beta)
+
+
+def try_descendant(S, C, R1_c_list, R2, N, alpha, x, beta):
+    """
+    It solves the problem of trying out each individual possibility for α^x.
+    """
+    D = C.copy()
+    A_dict = D.A_dict
+    if beta == D.n:
+        D.table.append([None]*len(D.A))
+        D.p.append(beta)
+    D.table[alpha][D.A_dict[x]] = beta
+    D.table[beta][D.A_dict_inv[x]] = alpha
+    D.deduction_stack.append((alpha, x))
+    D.process_deductions_check(R1_c_list[D.A_dict[x]], R1_c_list[D.A_dict_inv[x]])
+    if D.n == 0:
```

Comment:
Yes, that will be good from the implementation point of view. (though not from mathematician point of view, I believe). I agree this should be changed. :)